### ISSUE TITLE
Updating mods sometimes failes with beeing used error [windows]

### ISSUE THREAD

**Issue by knoxfighter** [mroote/factorio-server-manager#154]
The Mod-Update buttons sometimes result in an Internal Server Error. Reloading the page and try it again, will normally work. This also occures when clicking on the "Update All Mods" Button.

Console-Output:
```
2019/05/31 19:57:37 --------------------------------------------------------------
2019/05/31 19:57:38 download complete
 StatusCode: 200
 Status: 200 OK
2019/05/31 19:57:38 ModInfoList ... error when deleting mod: remove C:\Users\Knox\Desktop\Factorio\Factorio_0.17.45\mods/boblibrary_0.17.0.zip: The process cannot access the file because it is being used by another process.
2019/05/31 19:57:38 error when deleting mod: remove C:\Users\Knox\Desktop\Factorio\Factorio_0.17.45\mods/boblibrary_0.17.0.zip: The process cannot access the file because it is being used by another process.
2019/05/31 19:57:38 error when creating Mod: remove C:\Users\Knox\Desktop\Factorio\Factorio_0.17.45\mods/boblibrary_0.17.0.zip: The process cannot access the file because it is being used by another process.
2019/05/31 19:57:38 updateMod ... error when downloading the new Mod: remove C:\Users\Knox\Desktop\Factorio\Factorio_0.17.45\mods/boblibrary_0.17.0.zip: The process cannot access the file because it is being used by another process.
```

Also updated mods are there twice. The File with the old version and the file with the new version.

Seems like FSM blocks the access by itself.

**Comment by knoxfighter** [comment]
Also deleting sometimes fails!
It seams like the manager can not remove the specified file. When deliting this not even results in an error :(

**Comment by Psychomantis71** [comment]
@knoxfighter can you remove your personal data from the logs and post it here? I'll try replicating the error myself and perhaps fixing it.

Cheers!

**Comment by knoxfighter** [comment]
I already postet a small version, where the error is, but here is a full Log. It is very big, cause the FSM, is running for a month now.
Error occures in line 1250 and line 2101. In both sections, the error occures mutliple times.

[cleaned_fsm_log.log](https://github.com/mroote/factorio-server-manager/files/3437458/cleaned_fsm_log.log)

**Comment by Psychomantis71** [comment]
Thank you for the full log, I was expecting this to happen when the factorio server was up, shut down and then tried to update the mods (perhaps some process left behind still accessing the mods). But from the log I can see that you fully stopped FSM, clean start and it still happend, weird.

I'll replicate this both on windows server 2012 and centos and report back with my findings!

**Comment by knoxfighter** [comment]
I think, this is created, by the way, the files are locked by windows and my code.

I dont have problems with this, on Ubuntu 16.04 and Ubuntu 18.04, but on my Server (Windows Server 2016) the error occures. Semms like an Windows only error.

**Comment by Psychomantis71** [comment]
@knoxfighter I managed to replicate the error on windows no problem, but I am also getting a "zip: not a valid zip file" error. After I added in the code a loop for the program to wait until it can delete the file deletion went smooth (similar to the locking of the file). But still getting the zip error causing the program to crash fully... something is off with the file lock on windows for sure, will have to do more testing.

**Comment by knoxfighter** [comment]
Have in mind, that go has a different way to make concurency. It is not using different threads, with different PIDs, but it is using one thread, that works on different tasks, so they have the same PID. Thats why i added the extra library, for locking files with mutexes.

**Comment by Psychomantis71** [comment]
@knoxfighter did some testing, seems like the file is not being locked at all or incorrectly on windows. It will pass the LockW function without any errors, but upon the calling the os.remove it says "The process cannot access the file because it is being used by another program".
Perhaps we could try it with some other library?
Is this bug something that occurred before too?

**Comment by knoxfighter** [comment]
I think, it got introduced, when i rewrote the whole Mods Section and made it go-"Thread"-safe. I think, it is caused by the Windows-Kernel or golang on Windows.
We can try to also lock and unlock the files in vendor/lockfile.go, with the OS-sepcific way for filelocking.

**Comment by knoxfighter** [comment]
We can also look for a library, that also locks the file, for other processes and for this process.
Last time i looked, there was no such library, but i think, [flock](https://github.com/gofrs/flock) can do it. It just needs testing :)

**Comment by knoxfighter** [comment]
This is fixed in 0.8.2.
